src/main/java/com/amazon/opendistroforelasticsearch/sql/optimizer/subquery/SubqueryOptimizeRule.java

```diff
@@ -0,0 +1,76 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.optimizer.subquery;
+
+import com.alibaba.druid.sql.ast.expr.SQLQueryExpr;
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+import com.amazon.opendistroforelasticsearch.sql.optimizer.OptimizeRule;
+import com.amazon.opendistroforelasticsearch.sql.optimizer.subquery.InSubqueryRewriter;
+import com.amazon.opendistroforelasticsearch.sql.optimizer.subquery.model.Subquery;
+import com.amazon.opendistroforelasticsearch.sql.optimizer.subquery.model.SubqueryType;
+import com.amazon.opendistroforelasticsearch.sql.optimizer.subquery.rewriter.SubqueryAliasRewriter;
+import com.amazon.opendistroforelasticsearch.sql.optimizer.subquery.visitor.FindSubqueryInWhere;
+
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.Optional;
+
+/**
+ * Subquery Optimize Rule.
+ */
+public class SubqueryOptimizeRule implements OptimizeRule<SQLQueryExpr> {
+    private Optional<Subquery> subquery;
+
+    @Override
+    public boolean match(SQLQueryExpr expr) throws SQLFeatureNotSupportedException {
+        subquery = subquery(expr);
+        if (subquery.isPresent()) {
+            SubqueryType subqueryType = subquery.get().getSubqueryType();
+            if (subqueryType.isSupported()) {
+                return true;
+            } else {
+                throw new SQLFeatureNotSupportedException("Unsupported subquery. only IN is supported.");
```

Comment:
np: "only IN is supported" requires maintenance, ideally this message should be built using all the values except the unsupported.

Reply:
If the method returns a boolean result, to me the contract requires this case to just return false. I understand the intention, but I think the rule should not handle that error, some other higher level class should take care of that.